include/onnxruntime/core/session/onnxruntime_cxx_api.h

```diff
@@ -1508,6 +1512,49 @@ struct KernelContext {
   OrtKernelContext* ctx_;
 };
 
+// Logs a message using the provider Ort::Logger, which must be passed by value or reference.
+// Does not throw exceptions, so can be used in destructors and catch clauses.
+// Example: ORT_CXX_LOG(logger, ORT_LOGGING_LEVEL_INFO, "Log a message");
+#define ORT_CXX_LOG(logger, severity, message) \
+  std::ignore = logger.LogMessage(severity, __FILE__, __LINE__, static_cast<const char*>(__FUNCTION__), message)
+
+// Logs a message using the provider Ort::Logger, which must be passed by value or reference.
+// Throws an exception on failure.
+// Example: ORT_CXX_LOG_CHECKED(logger, ORT_LOGGING_LEVEL_INFO, "Log a message");
+#define ORT_CXX_LOG_CHECKED(logger, severity, message) \
+  Ort::ThrowOnError(logger.LogMessage(severity, __FILE__, __LINE__, static_cast<const char*>(__FUNCTION__), message))
+
+// Logs a formatted message using the provided Ort::Logger, which must be passed by value or reference.
+// Does not throw exceptions, so can be used in destructors and catch clauses.
+// Example: ORT_CXX_LOGF(logger, ORT_LOGGING_LEVEL_INFO, "Log an int: %d", 12);
+#define ORT_CXX_LOGF(logger, severity, ...) \
+  std::ignore = logger.LogFormattedMessage(severity, __FILE__, __LINE__, static_cast<const char*>(__FUNCTION__), __VA_ARGS__)
+
+// Logs a formatted message using the provided Ort::Logger, which must be passed by value or reference.
+// Throws an exception on failure.
+// Example: ORT_CXX_LOGF_CHECKED(logger, ORT_LOGGING_LEVEL_INFO, "Log an int: %d", 12);
+#define ORT_CXX_LOGF_CHECKED(logger, severity, ...) \
+  Ort::ThrowOnError(logger.LogFormattedMessage(severity, __FILE__, __LINE__, static_cast<const char*>(__FUNCTION__), __VA_ARGS__))
+
+struct Logger {
+  explicit Logger(const OrtLogger* logger);  ///< Uses an existing ::OrtLogger instance
+
+  OrtLoggingLevel GetLoggingSeverityLevel() const;  ///< Wraps OrtApi::Logger_GetLoggingSeverityLevel
+
+  ///< Wraps OrtApi::Logger_LogMessage. Typically called via the ORT_CXX_LOG or ORT_CXX_LOF_CHECHED macros.
+  Status LogMessage(OrtLoggingLevel log_severity_level, const char* file_path, int line_number,
+                    const char* func_name, const char* message) const noexcept;
+
+  ///< Prints a formatted message to a buffer and then calls OrtApi::Logger_LogMessage.
+  ///< Typically called via the ORT_CXX_LOGF or ORT_CXX_LOGF_CHECKED macros.
```

Comment:
We may add a note that the function may fail due to formatting error.

Reply:
Added